File: map-service/src/main/java/com/example/map/event/MarkerEventPublisher.java
package com.example.map.event;

import lombok.RequiredArgsConstructor;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class MarkerEventPublisher {

    private static final String TOPIC = "marker-created";

    private final KafkaTemplate<String, MarkerCreatedEvent> kafkaTemplate;

    public void publish(MarkerCreatedEvent event) {
        kafkaTemplate.send(TOPIC, event);
    }
}

File: apps/map-service/src/event/marker-event.publisher.ts
import { Injectable, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Kafka, Producer } from 'kafkajs';
import { MarkerCreatedEvent } from './marker-created.event';

@Injectable()
export class MarkerEventPublisher implements OnModuleInit, OnModuleDestroy {
  private readonly producer: Producer;

  constructor(private readonly config: ConfigService) {
    const kafka = new Kafka({
      brokers: [config.get('KAFKA_BROKERS', 'localhost:9092')],
    });
    this.producer = kafka.producer();
  }

  async onModuleInit(): Promise<void> {
    await this.producer.connect();
  }

  async onModuleDestroy(): Promise<void> {
    await this.producer.disconnect();
  }

  async publish(event: MarkerCreatedEvent): Promise<void> {
    await this.producer.send({
      topic: 'marker-created',
      messages: [{ value: JSON.stringify(event) }],
    });
  }
}
